### DISCUSSION TITLE
tm-xxx event not work on editor

### DISCUSSION THREAD

**Discussion by oeyoews** [TiddlyWiki/TiddlyWiki5#7956] [Q&A]
I hope to trigger some events through shortcut keys in the editor (cm6), or save the tiddler in real time. I tried the following code, but dispatchEvent does not seem to take effect. Only `tm-modal` takes effect. Did I go wrong? @Jermolene 

```js
    const title = $tw.wiki.getTiddler(widget?.editTitle!)?.fields[
      'draft.title'
    ] as string;
    const text = $tw.wiki.getTiddlerText(widget?.editTitle!);

    // NOTE: only update text, not include fields
    $tw.wiki.setText(title, 'text', '', text);
    $tw.notifier.display('saved');

    // not work
    // const childWidget = widget?.children[0].parentWidget!;
    // childWidget.dispatchEvent({
    //   type: 'tm-save-tiddler',
    //   param: title
    // });

    // not work still
    // $tw.rootWidget.invokeActionString();

    // work
    // $tw.rootWidget.dispatchEvent({ type: 'tm-modal', param: 'GettingStarted' });
```

**Comment by Jermolene** [comment]
Hi @oeyoews

One issue is that the `tm-save-tiddler` message has the same effect as clicking the tick button: not only does it save the draft tiddler, it replaces the editable draft with the underlying tiddler.

What are you trying to achieve?

**Reply by oeyoews** [reply]
just like ``<<save-tiddler-actions>>`` macro. 

```html
<$fieldmangler><$button tooltip={{$:/language/Buttons/Save-Open/Hint}} aria-label={{$:/language/Buttons/Save-Open/Caption}} class=<<tv-config-toolbar-class>>>
<<save-tiddler-actions>>
<$list filter="[all[current]get[draft.title]]">
<$action-sendmessage $message="tm-edit-tiddler"/>
</$list>
<$list filter="[<tv-config-toolbar-icons>prefix[yes]]">
{{$:/plugins/telmiger/EditButtons/images/done-open-button}}
</$list>
<$list filter="[<tv-config-toolbar-text>prefix[yes]]">
<span class="tc-btn-text"><$text text={{$:/language/Buttons/Save-Open/Caption}}/></span>
</$list>
</$button></$fieldmangler>
```

**Reply by Jermolene** [reply]
Hi @oeyoews are you trying to make a shortcut that saves a snapshot of the current draft back to the original tiddler?

**Comment by oeyoews** [comment]
yeah

**Reply by pmario** [reply]
IMO you should use `$tw.saverHandler.saveWiki()` as described at: https://talk.tiddlywiki.org/t/emergency-save-how-to-save-a-wiki-if-the-ui-is-broken/8656

Would that work for you?

```
$tw.saverHandler.saveWiki({template:"$:/core/save/all", 
method: "download",
downloadType: "text/plain", 
variables: {filename: "backup-x.html"}})
```

**Reply by Jermolene** [reply]
Hi @pmario I think @oeyoews wants a shortcut that will copy the current draft back over the original tiddler. It's the sort of thing that's handy when editing stylesheets or palettes and wants to see the effect of ones changes without leaving edit mode.

**Reply by oeyoews** [reply]
`telmiger/EditButtons` is a good plugin for it, so I hope it can be integrated into the cm6 plugin to implement some functions, such as saving or real-time automatic saving.

**Comment by oeyoews** [comment]
https://github.com/oeyoews/tiddlywiki-codemirror6/blob/cf3cd98e04ba549bc5ce6e6fea508664e9b0cb9c/src/tiddlywiki-codemirror-6/modules/keymap/keymap.ts#L27 Here is a simple implementation, just updating the editor text
